### PR TITLE
Report example of unanalyzed node in `NullableWalker`

### LINKED ISSUE #65056: Assertion in nullable walker (some nodes were not analyzed)
[Update:] We're using this bug to track making the assertion more actionable.

Message: 

> System.InvalidOperationException: 'Analyzed 58 nodes in NullableWalker, but DebugVerifier expects 62.'

Stack:

```
>	Microsoft.CodeAnalysis.Remote.ServiceHub.dll!Microsoft.CodeAnalysis.Remote.ThrowingTraceListener.Fail(string message, string detailMessage) Line 24	C#
 	System.dll!System.Diagnostics.TraceInternal.Fail(string message)	Unknown
 	System.dll!System.Diagnostics.Debug.Assert(bool condition, string message)	Unknown
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.NullableWalker.DebugVerifier.Verify(System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.CSharp.BoundExpression, (Microsoft.CodeAnalysis.NullabilityInfo Info, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol Type)> analyzedNullabilityMap, Microsoft.CodeAnalysis.CSharp.NullableWalker.SnapshotManager snapshotManagerOpt, Microsoft.CodeAnalysis.CSharp.BoundNode node) Line 54	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.NullableWalker.AnalyzeAndRewriteSpeculation(int position, Microsoft.CodeAnalysis.CSharp.BoundNode node, Microsoft.CodeAnalysis.CSharp.Binder binder, Microsoft.CodeAnalysis.CSharp.NullableWalker.SnapshotManager originalSnapshots, out Microsoft.CodeAnalysis.CSharp.NullableWalker.SnapshotManager newSnapshots, ref System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.CSharp.Symbol, Microsoft.CodeAnalysis.CSharp.Symbol> remappedSymbols) Line 1516	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.EnsureNullabilityAnalysisPerformedIfNecessary() Line 1955	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode node) Line 2052	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetLowerBoundNode(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode node) Line 544	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode node, out Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode bindableNode, out Microsoft.CodeAnalysis.CSharp.BoundNode lowestBoundNode, out Microsoft.CodeAnalysis.CSharp.BoundNode highestBoundNode, out Microsoft.CodeAnalysis.CSharp.BoundNode boundParent) Line 1328	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetSymbolInfoWorker(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode node, Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.SymbolInfoOptions options, System.Threading.CancellationToken cancellationToken) Line 1206	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.SpeculativeSemanticModelWithMemberModel.GetSymbolInfoWorker(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode node, Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.SymbolInfoOptions options, System.Threading.CancellationToken cancellationToken) Line 448	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfo(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, System.Threading.CancellationToken cancellationToken) Line 573	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfoFromNode(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) Line 4939	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfoCore(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) Line 5038	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.SemanticModel.GetSymbolInfo(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) Line 99	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.ModelExtensions.GetSymbolInfo(Microsoft.CodeAnalysis.SemanticModel semanticModel, Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) Line 21	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.IntroducedAmbiguity(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castNode, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax rewrittenExpression, Microsoft.CodeAnalysis.SemanticModel originalSemanticModel, Microsoft.CodeAnalysis.SemanticModel rewrittenSemanticModel, System.Threading.CancellationToken cancellationToken) Line 926	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.GetSemanticModelWithCastRemoved(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castNode, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castedExpressionNode, Microsoft.CodeAnalysis.SemanticModel originalSemanticModel, System.Threading.CancellationToken cancellationToken) Line 1289	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.IsConversionCastSafeToRemove(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castNode, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castedExpressionNode, Microsoft.CodeAnalysis.SemanticModel originalSemanticModel, Microsoft.CodeAnalysis.Operations.IConversionOperation originalConversionOperation, System.Threading.CancellationToken cancellationToken) Line 320	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.IsCastSafeToRemove(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castNode, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax castedExpressionNode, Microsoft.CodeAnalysis.SemanticModel originalSemanticModel, System.Threading.CancellationToken cancellationToken) Line 198	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.IsUnnecessaryCast(Microsoft.CodeAnalysis.CSharp.Syntax.CastExpressionSyntax cast, Microsoft.CodeAnalysis.SemanticModel semanticModel, System.Threading.CancellationToken cancellationToken) Line 108	C#
 	Microsoft.CodeAnalysis.CSharp.Workspaces.dll!Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.CastSimplifier.IsUnnecessaryCast(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax cast, Microsoft.CodeAnalysis.SemanticModel semanticModel, System.Threading.CancellationToken cancellationToken) Line 47	C#
 	Microsoft.CodeAnalysis.CSharp.Features.dll!Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer.IsUnnecessaryCast(Microsoft.CodeAnalysis.SemanticModel model, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax cast, System.Threading.CancellationToken cancellationToken) Line 27	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.RemoveUnnecessaryCast.AbstractRemoveUnnecessaryCastDiagnosticAnalyzer<Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax>.TryRemoveCastExpression(Microsoft.CodeAnalysis.SemanticModel model, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, System.Threading.CancellationToken cancellationToken) Line 57	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.RemoveUnnecessaryCast.AbstractRemoveUnnecessaryCastDiagnosticAnalyzer<Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax>.AnalyzeSyntax(Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext context) Line 42	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteSyntaxNodeAction.AnonymousMethod__62_0((System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext> action, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext context) data) Line 937	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock<(System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext)>(Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, System.Action<(System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext)> analyze, (System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext) argument, Microsoft.CodeAnalysis.Diagnostics.AnalysisContextInfo? info) Line 1620	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows<(System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext)>(Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, System.Action<(System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext)> analyze, (System.Action<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>, Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext) argument, Microsoft.CodeAnalysis.Diagnostics.AnalysisContextInfo? info) Line 1601	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteSyntaxNodeAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind>(Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalyzerAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind> syntaxNodeAction, Microsoft.CodeAnalysis.SyntaxNode node, Microsoft.CodeAnalysis.ISymbol containingSymbol, Microsoft.CodeAnalysis.SemanticModel semanticModel, System.Action<Microsoft.CodeAnalysis.Diagnostic> addDiagnostic, System.Func<Microsoft.CodeAnalysis.Diagnostic, bool> isSupportedDiagnostic, bool isGeneratedCode, Microsoft.CodeAnalysis.Diagnostics.AnalysisState.SyntaxNodeAnalyzerStateData analyzerState) Line 935	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteSyntaxNodeActions<Microsoft.CodeAnalysis.CSharp.SyntaxKind>(Microsoft.CodeAnalysis.SyntaxNode node, Microsoft.CodeAnalysis.Collections.ImmutableSegmentedDictionary<Microsoft.CodeAnalysis.CSharp.SyntaxKind, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalyzerAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind>>> nodeActionsByKind, Microsoft.CodeAnalysis.ISymbol containingSymbol, Microsoft.CodeAnalysis.SemanticModel model, System.Func<Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis.CSharp.SyntaxKind> getKind, System.Action<Microsoft.CodeAnalysis.Diagnostic> addDiagnostic, System.Func<Microsoft.CodeAnalysis.Diagnostic, bool> isSupportedDiagnostic, bool isGeneratedCode, Microsoft.CodeAnalysis.Diagnostics.AnalysisState.SyntaxNodeAnalyzerStateData analyzerState) Line 1418	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteSyntaxNodeActions<Microsoft.CodeAnalysis.CSharp.SyntaxKind>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> nodesToAnalyze, Microsoft.CodeAnalysis.Collections.ImmutableSegmentedDictionary<Microsoft.CodeAnalysis.CSharp.SyntaxKind, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalyzerAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind>>> nodeActionsByKind, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.ISymbol containingSymbol, Microsoft.CodeAnalysis.SemanticModel model, System.Func<Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis.CSharp.SyntaxKind> getKind, System.Action<Microsoft.CodeAnalysis.Diagnostic> addDiagnostic, System.Func<Microsoft.CodeAnalysis.Diagnostic, bool> isSupportedDiagnostic, bool isGeneratedCode, Microsoft.CodeAnalysis.Diagnostics.AnalysisState.SyntaxNodeAnalyzerStateData analyzerState) Line 1397	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteSyntaxNodeActionsCore<Microsoft.CodeAnalysis.CSharp.SyntaxKind>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> nodesToAnalyze, Microsoft.CodeAnalysis.Collections.ImmutableSegmentedDictionary<Microsoft.CodeAnalysis.CSharp.SyntaxKind, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalyzerAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind>>> nodeActionsByKind, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.ISymbol containingSymbol, Microsoft.CodeAnalysis.SemanticModel model, System.Func<Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis.CSharp.SyntaxKind> getKind, Microsoft.CodeAnalysis.Text.TextSpan filterSpan, Microsoft.CodeAnalysis.Diagnostics.AnalysisState.SyntaxNodeAnalyzerStateData analyzerState, bool isGeneratedCode) Line 1364	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.TryExecuteSyntaxNodeActions<Microsoft.CodeAnalysis.CSharp.SyntaxKind>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> nodesToAnalyze, Microsoft.CodeAnalysis.Collections.ImmutableSegmentedDictionary<Microsoft.CodeAnalysis.CSharp.SyntaxKind, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalyzerAction<Microsoft.CodeAnalysis.CSharp.SyntaxKind>>> nodeActionsByKind, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.SemanticModel model, System.Func<Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis.CSharp.SyntaxKind> getKind, Microsoft.CodeAnalysis.Text.TextSpan filterSpan, int declarationIndex, Microsoft.CodeAnalysis.ISymbol declaredSymbol, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope analysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState analysisState, bool isGeneratedCode) Line 1331	C#
```

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes #65056.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs [hunk 1]
```diff
@@ -46,12 +46,20 @@ public static void Verify(ImmutableDictionary<BoundExpression, (NullabilityInfo
...
+                        }
+                    }
+                }
+                else if (verifier._analyzedNullabilityMap.Count < verifier._visitedExpressions.Count)
+                {
+                    foreach (var verifiedNode in verifier._visitedExpressions)
+                    {
+                        if (!verifier._analyzedNullabilityMap.ContainsKey(verifiedNode))
+                        {
+                            Debug.Assert(false, $"Analyzed {verifier._analyzedNullabilityMap.Count} nodes in NullableWalker, but DebugVerifier expects {verifier._visitedExpressions.Count}. Example of unanalyzed node: {verifiedNode.GetDebuggerDisplay()}");
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs:59]: The one worry I have with this form is what happens in CI. If this happens locally we can easily attach a debugger, step through and see all the nodes. In CI though we throw on the first `Debug.Assert` and hence only get the message for the first node. It's possible the first node here turns out to be a symptom, but not the root of the problem. I wonder if it's worth changing this such that the assert contains all the nodes. 

Could approach this by just creating a `StringBuilder`, and everywhere we `Debug.Assert(false, ...` now we just append the message to the builder. Then at the end of `Verify` we assert if the `StringBuilder` isn't empty. 

Curious what others thoughts are here.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1). I hope that `GetDebuggerDisplay` (which is bound node type plus syntax if available) will be enough information. Let's see how this goes, we may have to output more information.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Comment by jcouv** [comment]
@333fred @RikkiGibson for a second review. Thanks

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs:59]: This might improve things, but I'm also happy to take the PR as-is.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**